### PR TITLE
build: Reduce intermediate layers

### PR DESCRIPTION
**Pull request by krishung5**
#### What does the PR do?

Reduce intermediate layers in order to reduce TRT-LLM container size.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [ ] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [x] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [x] build
- [ ] ci
- [ ] docs
- [ ] feat
- [ ] fix
- [ ] perf
- [ ] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Test plan:

TRT-LLM backend build + tests should pass.

- CI Pipeline ID: 16294063

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
- closes GitHub issue: #xxx

### CODE REVIEW

**Comment by nv-kmcgill53** [comment]
What is the improvement with this change?

**Comment by krishung5** [comment]
It seems like every Dockerfile command, e.g. FROM, , RUN, etc. creates an image layer of its own. The layers are then used to assemble the final image. With more accumulated layers, the image will be increased in size.

**Review by nv-kmcgill53 — APPROVED** [review]

File: build.py [hunk 1]
```diff
@@ -1082,25 +1082,20 @@ def create_dockerfile_linux(
...
+# Remove contents that are not needed in runtime
 # Setuptools has breaking changes in version 70.0.0, so fix it to 69.5.1
 # The generated code in grpc_service_pb2_grpc.py depends on grpcio>=1.64.0, so fix it to 1.64.0
-RUN pip3 install setuptools==69.5.1 grpcio-tools==1.64.0
+RUN ldconfig && \
+    ARCH="$(uname -i)" && \
+    rm -fr ${TRT_ROOT}/bin ${TRT_ROOT}/targets/${ARCH}-linux-gnu/bin ${TRT_ROOT}/data && \
+    rm -fr ${TRT_ROOT}/doc ${TRT_ROOT}/onnx_graphsurgeon ${TRT_ROOT}/python && \
+    rm -fr ${TRT_ROOT}/samples ${TRT_ROOT}/targets/${ARCH}-linux-gnu/samples && \
+    python3 -m pip install --upgrade pip && \
```
Comment (mc-nv) [inline build.py:1094]: I would rather keep `pip` to have  clean docker file.

File: build.py [hunk 2]
```diff
@@ -1082,25 +1082,20 @@ def create_dockerfile_linux(
...
-RUN find /usr -name libtensorrt_llm.so -exec dirname {} \; > /etc/ld.so.conf.d/tensorrt-llm.conf
-RUN find /opt/tritonserver -name libtritonserver.so -exec dirname {} \; > /etc/ld.so.conf.d/triton-tensorrtllm-worker.conf
-
+# Remove contents that are not needed in runtime
 # Setuptools has breaking changes in version 70.0.0, so fix it to 69.5.1
 # The generated code in grpc_service_pb2_grpc.py depends on grpcio>=1.64.0, so fix it to 1.64.0
-RUN pip3 install setuptools==69.5.1 grpcio-tools==1.64.0
+RUN ldconfig && \
+    ARCH="$(uname -i)" && \
+    rm -fr ${TRT_ROOT}/bin ${TRT_ROOT}/targets/${ARCH}-linux-gnu/bin ${TRT_ROOT}/data && \
```
Comment (mc-nv) [inline build.py:1091]: Double slash `\\` in the and is an escape symbol was to give developers more clean and readable `Dockerfile` output. Not sure if it make sense to remove it.

**Comment by mc-nv** [comment]
> It seems like every Dockerfile command, e.g. FROM, , RUN, etc. creates an image layer of its own. The layers are then used to assemble the final image. With more accumulated layers, the image will be increased in size.

Idea behind image layers is to keep layer history transparent. 
There why it's better to keep them separate unless it vital.

**Comment by krishung5** [comment]
@GuanLuo The container size from the latest pipeline is now only 24.7G, reducing by 8.7G.

**Comment by krishung5** [comment]
Will create a follow up PR to address @mc-nv's comments.

File: build.py (hunk 1 above)
Comment (krishung5) [reply build.py:1094]: Do you mean just run `pip install` instead of `python3 -m pip install`?

**Pull request merged**